### ISSUE TITLE
As <audio> replacement, plays twice then works no more

### ISSUE THREAD

**Issue by d01010101** [brion/ogv.js#469]
Hi, if OGVPlayer is used identically to an &lt;audio&gt; element, i.e. instead of document.getElementById() which looks up &lt;audio&gt;, a global javascript variable is used, initialized with OGVPlayer(), and the only methods called are .src(), .play(), .pause(), then OGVPlayer works incorrectly. The first issue happens only in Firefox:  OGVPlayer plays an ogg file twice after play(), where &lt;audio&gt; would play it once as expected. In Chrome both players play the stream once. The second issue happens in both Firefox and Chrome. OGVPlayer can be paused/replay several times, yet only before the stream reaches its end. Then play() works no more until the page is reloaded, while &lt;audio&gt; would play the same stream again any number of times.

**Comment by brion** [comment]
Can you provide a sample page?

**Comment by d01010101** [comment]
Hi Brion, I have a problem finding a host of ogv.js, your sample site does not allow linking from another site. But the bug can be reproduced for basic files like that:

<html><head>
<meta http-equiv="Content-Type" content="text/html;charset=UTF-8">
</head><body>
<script src="ogv/ogv.js"></script>
<script>
var audio_ogg = new OGVPlayer();
audio_ogg.src = "https://upload.wikimedia.org/wikipedia/commons/c/c8/Example.ogg";
audio_ogg.play();
</script>
</body>
</html>

It uses ogv.js-1.3.1 (the latest ZIP available for download) in a subdirectory ogvjs.

The problem is a bit different this time: Firefox sometimes asks it to stop a busy script, sometimes gets completely locked and takes up a lot of memory. Chrome in turn plays the example OGG as if it were a scratched vinyl plate as recorded here: http://picosong.com/pTHF/

**Comment by brion** [comment]
Hmm, copying that into an `index.html` and putting the 1.3.1 release into 'ogv' subdir works for me, at least on local machine. Firefox, Chrome, and Safari all play the file through to the end.

Check with git master at the demo page's minimal version: https://brionv.com/misc/ogv.js/demo/minimal.html

Open up JS console and type eg:

```
a = new OGVPlayer(); a.src = "https://upload.wikimedia.org/wikipedia/commons/c/c8/Example.ogg"; a.play();
```

For me, this works fine in Firefox; it plays to completion, and a second run of a.play() plays it a second time. In Chrome and Safari it runs fine until near the end, at which point it halts a fraction of a second early. Telling it to play again plays the last fraction of a second followed by a second run.

**Comment by brion** [comment]
Here's a copy of your simplified sample page with the 1.3.1 release, on a public host:
https://brionv.com/misc/ogv.js/bugtest/

Can you reproduce the bug there? If so, there might be a client-dependent issue. What OS are you running on?

**Comment by d01010101** [comment]
Yes, the behaviour is the same on https://brionv.com/misc/ogv.js/bugtest/

I guess I run a fairly standard configuration, no customized browsers etc.
```
4.4.0-53-generic #74-Ubuntu SMP Fri Dec 2 15:59:10 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
Firefox 50.0.2
56.0.2924.87 (64-bit)
```

Here is a console log of https://brionv.com/misc/ogv.js/bugtest in Chrome:

ogv.js:6553 The provided value 'moz-chunked-arraybuffer' is not a valid enum value of type XMLHttpRequestResponseType.
MozChunkedBackend.supported @ ogv.js:6553
ogv.js:7129 The provided value 'ms-stream' is not a valid enum value of type XMLHttpRequestResponseType.
MSStreamBackend.supported @ ogv.js:7129
ogv.js:9223 slow demux in 14.1400000000001 ms; 0 +video packets, 2 +audio packets
ogv.js:9223 slow demux in 453.6550000000001 ms; 0 +video packets, 0 +audio packets
ogv.js:9223 slow demux in 55.649999999999864 ms; 0 +video packets, -1 +audio packets

**Comment by brion** [comment]
Curious, I'll test under Linux tomorrow when my machines are free for a reboot. Should be working, but there might be surprises in the audio configuration that could affect behavior...

The console output looks okay; the warnings about the response type are from the autodetection of available data-fetching modes, and the slow demux is a debug message that can be ignored.

**Comment by d01010101** [comment]
Might be audio configuration, but your other demos, both audio and audio+video, work flawlessly on the same system.

**Comment by d01010101** [comment]
Might be related or not, but on Safari there is a short repeated part when played pause. Again, the problem concers the "invisible" version controlled like in https://brionv.com/misc/ogv.js/bugtest.

**Comment by d01010101** [comment]
Hi Brion, did you test the issue on other Linuxes? If AAC is paused on Safari, a bit of sound is skipped, OGV in turn plays a small fragment of OGG twice (players without own controls). There is no way I know of which would circumvent these problems.

**Comment by d01010101** [comment]
Now it plays correctly in Firefox, still stuttering in Chrome, even if the library is of the same version, so it seems it is a random bug. When it stutters, the script loops making a page unresponsive and this shows every few seconds:

slow demux in 17.70499999999811 ms; 0 +video packets, 0 +audio packets
ogv.js?version=OGV_VERSION:9223 slow demux in 8.775000000001455 ms; 0 +video packets, 0 +audio packets
ogv.js?version=OGV_VERSION:9223 slow demux in 8.784999999996217 ms; 0 +video packets, 0 +audio packets
ogv.js?version=OGV_VERSION:9223 slow demux in 15.430000000000291 ms; 0 +video packets, 0 +audio packets
ogv.js?version=OGV_VERSION:9223 slow demux in 17.99000000000524 ms; 0 +video packets, 0 +audio packets
ogv.js?version=OGV_VERSION:9223 slow demux in 13.74500000000262 ms; 0 +video packets, 0 +audio packets

The packet values are always 0 when in this loop. If it were any help, I can e.g. debug a version with diagnostics on my browser.

On Android the library works correctly.

**Comment by d01010101** [comment]
My last debug attempt, so that I won't flood the developers in case they choose to ignore the issue.

On the same browser (Linux/Chrome) https://brionv.com/misc/ogv.js/demo/ plays both OGV and OGG without problems, play/pause etc., yet https://brionv.com/misc/ogv.js/bugtest falls into an infinite loop on the same browser, ovg.js versions tested 1.1.2 ... 1.4.1.

**Comment by d01010101** [comment]
Infitive loop - clearing browser cache helped, but ogv has probably some issue with playing truncated/corrupted files.

There is no support for event listeners probably. Using multiple players per page gives weird problems like playing a file twice with a bit of noise in between. There are different issues in different browsers like not always reacts on pause, truncates a file a second early. I can't exclude problems in my app, still I highly doubt that it would be able to force ogv to play a file twice especially if there is no `ended' handler, to produce random noise and to end a file a second early.

This thread changed into general rambling instead of a bug report, so it should be closed, but
anyway, ogv does not seem unfortunately to be a mature technology yet, with random bug browsers and incompatibities not helping.

**Comment by brion** [comment]
Sorry it's not working out for you. :( Unfortunately I haven't been able to successfully reproduce the stuttering in Chrome under Linux (Fedora or Ubuntu), or the infinite loop (though I had seen some such problems on an older version, believed to be fixed...) so haven't been able to diagnose any detail issues there.

Note that in general I wouldn't recommend using ogv.js when native Ogg or WebM support is available in &lt;audio>, but it *should* work reliably under Linux as well as other OSs and I've had no Linux-specific problems locally, so I'd like to work them out if possible.

The early truncation bug should be fixed in 1.4.x -- https://github.com/brion/ogv.js/commit/6f34981aa12f25556b32015fda085a58581c90b5

If you have problems with specific event listeners being missing let me know -- not everything from HTMLMediaElement is implemented, so there may still be holes in the API.

> If AAC is paused on Safari, a bit of sound is skipped, OGV in turn plays a small fragment of OGG twice (players without own controls). There is no way I know of which would circumvent these problems.

Repetition of a small chunk of audio after pause/play in the middle of playback is probably related to https://github.com/brion/ogv.js/issues/458
